InfiniTAM/Engine/RosEngine.cpp

```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
 Vector2i RosEngine::getDepthImageSize(void) { return image_size_depth_; }
 Vector2i RosEngine::getRGBImageSize(void) { return image_size_rgb_; }
 
+sensor_msgs::PointCloud2 RosEngine::conversionToPCL() {
+  ROS_INFO("conversionToPCL() start");
+  pcl::PointCloud<pcl::PointXYZ> complete_point_cloud;
+  sensor_msgs::PointCloud2 complete_point_cloud2;
+
+  if (main_engine_ != NULL) {
+
+    CHECK_NOTNULL(main_engine_->GetMesh());
+
+    main_engine_->GetMeshingEngine()->MeshScene(main_engine_->GetMesh(),
+                                           main_engine_->GetScene());
+
+    ORUtils::MemoryBlock<ITMMesh::Triangle>* cpu_triangles;
+    bool shoulDelete = false;
+
+    if (main_engine_->GetMesh()->memoryType == MEMORYDEVICE_CUDA) {
+      cpu_triangles = new ORUtils::MemoryBlock<ITMMesh::Triangle>(
+          main_engine_->GetMesh()->noMaxTriangles, MEMORYDEVICE_CPU);
+      cpu_triangles->SetFrom(
+          main_engine_->GetMesh()->triangles,
+          ORUtils::MemoryBlock<ITMMesh::Triangle>::CUDA_TO_CPU);
+      shoulDelete = true;
+    } else {
+      cpu_triangles = main_engine_->GetMesh()->triangles;
+    }
+
+    ITMMesh::Triangle* triangleArray = cpu_triangles->GetData(MEMORYDEVICE_CPU);
+
+    complete_point_cloud.width =
+        main_engine_->GetMesh()->noTotalTriangles * 3; // Point cloud has 3 points per triangle.
```

Comment:
Thanks, I meant the triangle has at least 3 points, not the point cloud. Now what do you advise, check if there are duplicates in the point cloud (cpu expensive?) or let the point cloud through a voxel grid filter at the end of the method? 

I tried to figure out how STL files are build and if PCL has a method to transform from mesh to pcl. They have but as far as I can see they also do not count for the fact that there could be duplicates in the point cloud (https://github.com/PointCloudLibrary/pcl/blob/46cb8fe5589e88e36d79f9b8b8e5f4ff4fceb5de/io/src/vtk_lib_io.cpp )